node/core/dispute-coordinator/src/real/initialized.rs

```diff
@@ -832,7 +832,8 @@ impl Initialized {
 			.is_some();
 
 		// Update candidate votes.
-		for (statement, val_index) in &statements {
+		let mut fresh_statements = HashSet::new(); // Used to keep track if the statement is fresh or not. Needed for spam detection.
+		for (idx, (statement, val_index)) in statements.iter().enumerate() {
```

Comment:
Why don't we just simply consume statements here via `into_iter` and push the whole statement data into a `fresh_invalid_statements` vec directly, in case it was found to be fresh & invalid. This way we can avoid the indexing - which is always good as it provides room for out of bounds errors. In the loop below we can drop the check for validity then, as the vec will only contain the invalid votes to begin with.

Reply:
I agree, indexing is error prone and I didn't realise we care only about invalid statements.

This however means that we don't need to keep the invalid statements at all, because we only use the validator index (the actual index - `val_index`, not the index in the array). So bottom line - we can push only `val_index` in the vector.

Am I misunderstanding something?

Reply:
You are right, we only need the `ValidatorIndex` - let's just keep that.